Azure - VM , SQLdb, Load Balancer , Virtual Networks and CDN

Thushara67 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Azure cloudinventory

Added azurelib package which helps in gathering info about vms and sql dbs for a subscriptionID.
This package contains vms.go,dbs.go and base.go which includes function to get all subscriptionIDs in an account level
This package also contains unit tests azure_test.go and base_test.go

Added azurecollect package which helps in gathering info about sqldbs for subscriptionIDs in an account level.

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Thushara67 committed:
added azurelib and azure collector packages

Thushara67 committed:
added azurelib and azure collector packages

Thushara67 committed:
updated dbs.go

Thushara67 committed:
updated dbs.go

ManishKumarBachhu committed:
Updated GetAllVMS

Changed parameters to subscriptionID and added clients and context inside the function.

ManishKumarBachhu committed:
Updated GetAllVMS

Changed parameters to subscriptionID and added clients and context inside the function.

ManishKumarBachhu committed:
Added CollectVMS function

A collector function for gathering virtual machines across all subscriptions.

ManishKumarBachhu committed:
Added CollectVMS function

A collector function for gathering virtual machines across all subscriptions.

Thushara67 committed:
Update vms.go

Thushara67 committed:
Update vms.go

Thushara67 committed:
Update azure_test.go

Thushara67 committed:
Update azure_test.go

Thushara67 committed:
Update azurecollect.go

Thushara67 committed:
Update azurecollect.go

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//VirtualMachineinfo is a  struct  that contains information related to a virtual machine
+type VirtualMachineinfo struct {
```
VirtualMachineInfo would be the correct camel-case.
Please change the name

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+//VirtualMachineinfo is a  struct  that contains information related to a virtual machine
+type VirtualMachineinfo struct {
+        VM                  *compute.VirtualMachine
+        PrivateIpaddress    *string
```
I'll prefer a string here rather than a pointer.
Also, consistent camel casing. `IPAddress` would be more appropriate

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+//VirtualMachineinfo is a  struct  that contains information related to a virtual machine
+type VirtualMachineinfo struct {
+        VM                  *compute.VirtualMachine
+        PrivateIpaddress    *string
+        PublicIpname        *string
+        PublicIpaddress     *string
+        VirtualnetandSubnet *string
+        Ipconfig            *string
```
I'll prefer a string here rather than a pointer.
Also, consistent camel casing. `IPAddress` would be more appropriate

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+        "time"
+)
+
+//Clients is a struct that contains all the necessary clients
```
Please check all comments: 
It is idiomatic to have a whitespace after //
GoFmt/lint should already pick this up. I'm not sure though

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+// AuthorizeClients function authorizes all the clients
+func AuthorizeClients(c Clients) (Clients, error) {
```
Let's make this a method on the Clients struct
`func (c *Clients) Authorize() error`

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetallVMS function returns list of virtual machines
+func GetallVMS(subscriptionID string) (Vmlist []*VirtualMachineinfo, err error) {
```
`GetAllVMs`
`VMList`

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+//GetallVMS function returns list of virtual machines
+func GetallVMS(subscriptionID string) (Vmlist []*VirtualMachineinfo, err error) {
+
+        cl := GetNewClients(subscriptionID)
```
Make this an explicit dependency of the function.
Change the signature to require clients.
Also, as a general rule, only the clients that are actually required, should be added. i.e no need to pass all 3 clients if 1 will do the job.

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+                return nil, err
+        }
+
+        instancesChan := make(chan *VirtualMachineinfo, 1000)
```
I think this warrants a bigger discussion. I'll bring it up separately.
In general though, I don't think we need to wait till ALL details have been gathered to start emptying the channel. The main routine can start emptying the channel once all go routines have been fired. That way any buffer will work.

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetVMResourcegroup function returns resourcegroup to which the virtual machine belongs to
+func GetVMResourcegroup(vm *compute.VirtualMachine) (resourceGroup string, err error) {
```
`VMResourceGroup`

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetVMname function returns the virtual machine's name
+func GetVMname(vm *compute.VirtualMachine) (Name string, err error) {
```
`VMName`

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetPrivateIP function returns Private IP Address of a Virtual Machine
+func GetPrivateIP(ctx context.Context, client Clients,
```
if only Interface client is needed, change the signature to need only the one client.

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetPublicIPAddressID function returns Public IP Address ID (PublicIPName)
+func GetPublicIPAddressID(ctx context.Context,
```
if only Interface client is needed, change the signature to need only the one client.

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetPublicIPAddress function returns the PublicIPAddress of the virtual machine
+func GetPublicIPAddress(ctx context.Context, client Clients,
```
if only Interface client is needed, change the signature to need only the one client.

tchaudhry91 commented on azurelib/vms.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+//GetDNS function returns  DNS's Fqdn
+func GetDNS(ctx context.Context, client Clients,
```
if only Interface client is needed, change the signature to need only the one client.

tchaudhry91 commented on azurelib/dbs.go:
```diff
@@ -0,0 +1,62 @@
+                if err != nil {
+                        return
+                }
+                instancesChan := make(chan *sql.Database, 800)
```
Same concept as VMs.
Let's discuss this separately.

tchaudhry91 commented on azurelib/base_test.go:
```diff
@@ -0,0 +1,22 @@
+
+
+//TestGetallSubscriptionIDs tests function GetallSubscriptionIDs
+func TestGetallSubscriptionIDs(t *testing.T) {
```
`GetAll`

tchaudhry91 commented on azurelib/base.go:
```diff
@@ -0,0 +1,31 @@
+
+//GetallSubscriptionIDs function returns a map of subscription name and subscription ID at account level
+func GetallSubscriptionIDs(ctx context.Context) (map[string]string, error) {
+        subscriptionmap := make(map[string]string)
```
subscriptionMap

tchaudhry91 commented:
Overall, great start.
Added some in-line comments. The review is not yet complete, but let's start with the changes I proposed. Let's rinse and repeat a few times. 
General:
- Camel case needs to be applied properly. I spotted plenty of discrepancies. Please correct them and any others I may have missed.
- Leave a single whitespace after `//` in the comments. That's the idiomatic way, the formatter/linter should already have taken care of this, but I'm not sure what's wrong.

tchaudhry91 reviewed (changes requested):
Please re-request after changes

Thushara67 committed:
Update vms.go

Added camel casing to variables and added changes to GetALLVMS

Thushara67 committed:
Update vms.go

Added camel casing to variables and added changes to GetALLVMS

Thushara67 committed:
Update azure_test.go

Added camel casing and added changes to TestGetAllVMS according to changes made in GetAllVMS

Thushara67 committed:
Update azure_test.go

Added camel casing and added changes to TestGetAllVMS according to changes made in GetAllVMS

Thushara67 committed:
updated base.go

Added camel casing

Thushara67 committed:
updated base.go

Added camel casing

Thushara67 committed:
updated base_test.go

Added camel casing

Thushara67 committed:
updated base_test.go

Added camel casing

Thushara67 committed:
updated dbs.go

Added camel casing to variables and functions

Thushara67 committed:
updated dbs.go

Added camel casing to variables and functions

Thushara67 committed:
updated azurecollect.go

Added camel casing and added changes to CollectVMsPerSubscriptionID based on changes made in GetALLVMS

Thushara67 committed:
updated azurecollect.go

Added camel casing and added changes to CollectVMsPerSubscriptionID based on changes made in GetALLVMS

Thushara67 committed:
updated azurecollect.go

Added a function NewAzureCollector which returns a AzureCollector with subscriptionMap, populated with the subscription name and ID

Thushara67 committed:
updated azurecollect.go

Added a function NewAzureCollector which returns a AzureCollector with subscriptionMap, populated with the subscription name and ID

Thushara67 committed:
Added azurecollect_test.go

Thushara67 committed:
Added azurecollect_test.go

Thushara67 committed:
Merge branch 'azure' of https://github.com/adobe/cloudinventory into azure

Thushara67 committed:
Added azure.go 

azure.go contains azureCmd which is subcommand of dump.

Thushara67 committed:
Update dump.go

ManishKumarBachhu committed:
Changed VirtualMachineInfo structure

Added additional_info structure into VirtualMachineInfo

Thushara67 committed:
imported package azurelib

Thushara67 committed:
imported package azurecollector

Thushara67 committed:
Added loadbalancer.go

loadbalancer.go contains GetAllLdb function which helps in gathering information about all loadbalancers in a subscriptionID

Thushara67 committed:
Added  TestGetAllLdb function to azure_test.go

TestGetAllLdb function tests the function GetAllLdb

Thushara67 committed:
Added changes to concurrency implementation to GetAllVMS

Made changes in implementation of concurrency across vms for gathering their info

Thushara67 committed:
Added  SQLDBInfo struct 

SQLDBInfo struct has BasicInfo which is a struct returned by the sdk and AdditionalInfo which contains resourceGroup and serverName

Thushara67 committed:
Added changes to CollectSQLDBs to accomodate the changes made in GetAllSQLDBs in azurelib

Thushara67 committed:
removed unused package

ManishKumarBachhu committed:
Added Collector for Load Balancers

ManishKumarBachhu committed:
Added TestCollectLoadBalancers

Added TestCollectLoadBalancers() to test CollectLoadBalancers()

Thushara67 committed:
Added azure-sdk-for-go network package

Thushara67 committed:
added small change to error statement

Thushara67 committed:
Added filter option for load balancer

Thushara67 committed:
Added subscriptionID to additional info

ManishKumarBachhu committed:
Added SubscriptionID to structs

Added SubscriptionID  to Client and AdditionalInfo structs

Thushara67 committed:
Added  NewAzureCollectorUserDefined   function 

NewAzureCollectorUserDefined   function takes a list of subscriptionIDs and  returns an AzureCollector with subscription info given by user in subscriptionMap.

Thushara67 committed:
Added flag "inputPath" which takes subscriptionIDs from user through a file

Thushara67 committed:
Updated go.mod

Thushara67 committed:
Updated go.sum

Thushara67 committed:
Update azure.go

ManishKumarBachhu committed:
Create go.yml

Thushara67 committed:
Added loadbalancer in description for filter flag

Thushara67 committed:
Update README.md

Thushara67 committed:
Added maxGoRoutines as input to CollectVMS 

User can give max no.of  Go routines to be used to collect the information

Thushara67 committed:
Added flag -m  to specify max  no.of Goroutines

Thushara67 committed:
Added changes to TestCollectVMS

Thushara67 committed:
Added maxGoRoutines as input to collector function

Added maxGoRoutines as input to CollectLoadBalancers and CollectSQLDBs

Thushara67 committed:
Added changes to unit tests

Added changes to TestCollectSQLDBs,TestCollectLoadBalancer

Thushara67 committed:
Added maxGoRoutines as input to collector function

Added maxGoRoutines as input to collector function used in cmd

Thushara67 committed:
Added GetVMCount function

 GetVMCount function returns stats of virtual machines for a subscriptionID

Thushara67 committed:
Added GetSQLDBCount  function

GetSQLDBCount function returns stats of SQL databases for a given subscriptionID

Thushara67 committed:
Added  CollectVMSCount, CollectSQLDBCount  functions

Thushara67 committed:
Added changes to TestCollectLoadBalancers

Thushara67 committed:
Added flag -s to dump the stats of different resources for subscriptions

Thushara67 committed:
Added GetAllCDN function

GetAllCDN function returns a list of CDN instances for a given subscriptionID

Thushara67 committed:
Added CollectCDN function

CollectCDN gathers information and stats of CDN for each subscriptionID in an account level

Thushara67 committed:
Added filter option for CDN

Thushara67 committed:
Added TestGetAllCDN function

TestGetAllCDN tests the function GetAllCDN

Thushara67 committed:
Added TestCollectCDN function

TestCollectCDN test the function CollectCDN

Thushara67 committed:
Added cdn in description for filter flag

Thushara67 committed:
Update README.md
